### ISSUE TITLE
h5py unable to create file in gcs bucket.

### ISSUE THREAD

**Issue by h4ck4l1** [h5py/h5py#2375]
To assist reproducing bugs, please include the following:
 * Operating System (e.g. Windows 10, MacOS 10.11, Ubuntu 20.04.2 LTS, CentOS 7)

```
Ubuntu 22.04.3 LTS (Jammy Jellyfish) 
Google Colab
```
 * Python version (e.g. 3.10)

```
3.10.12 (Google Colab)
```

 * Where Python was acquired (e.g. system Python on MacOS or Linux, Anaconda on
   Windows)

```
Linux (Google Colab)
```

 * h5py version (e.g. 3.9)

```
3.9.0
```

 * HDF5 version (e.g. 1.12.2)

```
1.12.2
```

 * The full traceback/stack trace shown (if it appears)

```
Trial 2 Complete [00h 00m 45s]

Best Validation Accuracy So Far: None
Total elapsed time: 00h 02m 02s

Search: Running Trial #3

Value             |Best Value So Far |Hyperparameter
500               |300               |first_dense_units
100               |100               |second_dense_units
100               |50                |third_dense_units
0.001             |1e-05             |l_rate
2                 |2                 |tuner/epochs
0                 |0                 |tuner/initial_epoch
6                 |6                 |tuner/bracket
0                 |0                 |tuner/round

EPOCH: 1/2
19680/19680 ━━━━━━━━━━━━━━━━━━━━ 27s 1ms/step - Training Loss: 0.6296 - Training Accuracy: 0.7751
Total training loss for epoch 0: 0.6295972466468811
Total training metric for epoch 0: 0.7751016020774841
3968/3968 ━━━━━━━━━━━━━━━━━━━━ 33s 8ms/step - Validation Loss: 1.7660 - Validation Accuracy: 0.3737
Total validation loss for epoch 0: 1.7659825086593628
Total validation metric for epoch 0: 0.37373992800712585
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py", line 273, in _try_run_and_update_trial
    self._run_and_update_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py", line 238, in _run_and_update_trial
    results = self.run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/tuners/hyperband.py", line 427, in run_trial
    return super().run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py", line 314, in run_trial
    obj_value = self._build_and_fit_model(trial, *args, **copied_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py", line 233, in _build_and_fit_model
    results = self.hypermodel.fit(hp, model, *args, **kwargs)
  File "<ipython-input-16-63d28ced77c2>", line 110, in fit
    callback.on_epoch_end(epoch,logs={"Validation Accuracy":validation_accuracy})
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner_utils.py", line 76, in on_epoch_end
    self._save_model()
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner_utils.py", line 86, in _save_model
    self.model.save_weights(write_filepath)
  File "/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py", line 123, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/usr/local/lib/python3.10/dist-packages/h5py/_hl/files.py", line 567, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)
  File "/usr/local/lib/python3.10/dist-packages/h5py/_hl/files.py", line 237, in make_fid
    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 126, in h5py.h5f.create
FileNotFoundError: [Errno 2] Unable to create file (unable to open file: name = 'gs://stanfordrna/hyperband/mnist_hyperband/trial_0002/checkpoint.weights.h5', errno = 2, error message = 'No such file or directory', flags = 13, o_flags = 242)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-16-63d28ced77c2>](https://localhost:8080/#) in <cell line: 140>()
    138 train_ds = strategy.distribute_datasets_from_function(lambda _: create_train_ds(train_raw,per_sample_batch_size))
    139 valid_ds = strategy.distribute_datasets_from_function(lambda _: create_valid_ds(valid_raw,per_sample_batch_size))
--> 140 tuner.search(train_data=train_ds,validation_data=valid_ds,epochs=10,batch_size=BATCH_SIZE,steps_per_epoch=steps_per_epoch,validation_steps=validation_steps)

4 frames
[/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py](https://localhost:8080/#) in search(self, *fit_args, **fit_kwargs)
    232             self.on_trial_begin(trial)
    233             self._try_run_and_update_trial(trial, *fit_args, **fit_kwargs)
--> 234             self.on_trial_end(trial)
    235         self.on_search_end()
    236 

[/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py](https://localhost:8080/#) in on_trial_end(self, trial)
    336             trial: A `Trial` instance.
    337         """
--> 338         self.oracle.end_trial(trial)
    339         self.save()
    340 

[/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/oracle.py](https://localhost:8080/#) in wrapped_func(*args, **kwargs)
    106             LOCKS[oracle].acquire()
    107             THREADS[oracle] = thread_name
--> 108         ret_val = func(*args, **kwargs)
    109         if need_acquire:
    110             THREADS[oracle] = None

[/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/oracle.py](https://localhost:8080/#) in end_trial(self, trial)
    584         if not self._retry(trial):
    585             self.end_order.append(trial.trial_id)
--> 586             self._check_consecutive_failures()
    587 
    588         self._save_trial(trial)

[/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/oracle.py](https://localhost:8080/#) in _check_consecutive_failures(self)
    541                 consecutive_failures = 0
    542             if consecutive_failures == self.max_consecutive_failed_trials:
--> 543                 raise RuntimeError(
    544                     "Number of consecutive failures exceeded the limit "
    545                     f"of {self.max_consecutive_failed_trials}.\n"

RuntimeError: Number of consecutive failures exceeded the limit of 3.
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py", line 273, in _try_run_and_update_trial
    self._run_and_update_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/base_tuner.py", line 238, in _run_and_update_trial
    results = self.run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/tuners/hyperband.py", line 427, in run_trial
    return super().run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py", line 314, in run_trial
    obj_value = self._build_and_fit_model(trial, *args, **copied_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner.py", line 233, in _build_and_fit_model
    results = self.hypermodel.fit(hp, model, *args, **kwargs)
  File "<ipython-input-16-63d28ced77c2>", line 110, in fit
    callback.on_epoch_end(epoch,logs={"Validation Accuracy":validation_accuracy})
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner_utils.py", line 76, in on_epoch_end
    self._save_model()
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/src/engine/tuner_utils.py", line 86, in _save_model
    self.model.save_weights(write_filepath)
  File "/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py", line 123, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/usr/local/lib/python3.10/dist-packages/h5py/_hl/files.py", line 567, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)
  File "/usr/local/lib/python3.10/dist-packages/h5py/_hl/files.py", line 237, in make_fid
    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 126, in h5py.h5f.create
FileNotFoundError: [Errno 2] Unable to create file (unable to open file: name = 'gs://stanfordrna/hyperband/mnist_hyperband/trial_0002/checkpoint.weights.h5', errno = 2, error message = 'No such file or directory', flags = 13, o_flags = 242)
```

* Additional comments

```
The storage bucket exists I can post the gsutil outputs. I am using the owner account of the gcs which has all the permissions needed.
After starting the keras tuner search. The tuner is able to create folder and trail sub folders. but It fails saying It can't create file(?). 
And also yes I have raised this issue in keras tuner too and I am yet to get a reply from them I am raising this issue here because I thought maybe its related to h5py. 
If its not related to h5py please confirm and I will delete this issue
Thankyou.
```

**Comment by h4ck4l1** [comment]
* Relevant code input 

```
class CustomHypermodel(kt.HyperModel):

    def build(self,hp):
        with strategy.scope():
            model = Mnist(
            first_dense_units=hp.Int(name="first_dense_units",min_value=300,max_value=500,step=100),
            second_dense_units=hp.Int(name="second_dense_units",min_value=100,max_value=300,step=100),
            third_dense_units=hp.Int(name="third_dense_units",min_value=50,max_value=150,step=50)
            )
        return model

    def fit(self,
            hp,
            model,
            train_data,
            validation_data,
            epochs:Optional[int]=None,
            batch_size:Optional[int]=None,
            steps_per_epoch:Optional[int]=None,
            validation_steps:Optional[int]=None,
            callbacks:Optional[list[keras.callbacks.Callback]]=None,
            **kwargs):

        with strategy.scope():
            optimizer = keras.optimizers.Adam(learning_rate=hp.Float("l_rate",min_value=1e-5,max_value=1e-2,sampling="log",step=10,default=1e-5))
            training_loss = keras.metrics.Mean(name="train_loss",dtype=tf.float32)
            training_metric = keras.metrics.SparseCategoricalAccuracy(name="training_accuracy",dtype=tf.float32)
            validation_loss = keras.metrics.Mean(name="valid_loss",dtype=tf.float32)
            validation_metric = keras.metrics.SparseCategoricalAccuracy(name="validation_accuracy",dtype=tf.float32)



        @tf.function
        def train_step(iterator):

            def train_step_fn(inputs):

                images,y_true = inputs

                with tf.GradientTape() as tape:

                    y_pred = model(images,training=True)
                    per_example_loss = keras.losses.sparse_categorical_crossentropy(y_true,y_pred,from_logits=True)
                    loss = tf.nn.compute_average_loss(per_example_loss)

                grads = tape.gradient(loss,model.trainable_variables)
                optimizer.apply_gradients(list(zip(grads,model.trainable_variables)))
                training_loss.update_state(loss*strategy.num_replicas_in_sync)
                training_metric.update_state(y_true,y_pred)

            strategy.run(train_step_fn,args=(next(iterator),))



        @tf.function
        def valid_step(iterator):

            def valid_step_fn(inputs):

                images,y_true = inputs
                y_pred = model(images)
                loss = keras.losses.sparse_categorical_crossentropy(y_true,y_pred,from_logits=True)
                validation_loss.update_state(loss*strategy.num_replicas_in_sync)
                validation_metric.update_state(y_true,y_pred)


            strategy.run(valid_step_fn,args=(next(iterator),))


        for callback in callbacks:
            callback.set_model(model)


        best_validation_accuracy = 0

        train_iterator = iter(train_data)
        valid_iterator = iter(validation_data)

        for epoch in range(epochs):
            print(f"EPOCH: {epoch+1}/{epochs}")

            train_prog_bar = Progbar(steps_per_epoch * batch_size,stateful_metrics=["Training Loss","Training Accuracy"])
            valid_prog_bar = Progbar(validation_steps * batch_size,stateful_metrics=["Validation Loss","Validation Accuracy"])

            for step in range(steps_per_epoch):

                train_step(train_iterator)

                values = [("Training Loss",training_loss.result().numpy()),("Training Accuracy",training_metric.result().numpy())]
                train_prog_bar.add(batch_size,values=values)

            print(f"Total training loss for epoch {epoch}: {training_loss.result().numpy()}")
            print(f"Total training metric for epoch {epoch}: {training_metric.result().numpy()}")
            training_loss.reset_state()
            training_metric.reset_state()

            for step in range(validation_steps):

                valid_step(valid_iterator)

                values = [("Validation Loss",validation_loss.result().numpy()),("Validation Accuracy",validation_metric.result().numpy())]
                valid_prog_bar.add(batch_size,values=values)

            print(f"Total validation loss for epoch {epoch}: {validation_loss.result().numpy()}")
            print(f"Total validation metric for epoch {epoch}: {validation_metric.result().numpy()}")

            validation_accuracy = float(validation_metric.result().numpy())

            for callback in callbacks:
                callback.on_epoch_end(epoch,logs={"Validation Accuracy":validation_accuracy})

            validation_loss.reset_state()
            validation_metric.reset_state()


            best_validation_accuracy = max(best_validation_accuracy,validation_accuracy)


        return best_validation_accuracy




tuner = kt.Hyperband(
    objective=kt.Objective("Validation Accuracy","max"),
    max_epochs=100,
    factor=2,
    hypermodel=CustomHypermodel(),
    hyperband_iterations=2,
    directory="gs://stanfordrna/hyperband",
    project_name="mnist_hyperband",
    overwrite=True
)



per_sample_batch_size = BATCH_SIZE // strategy.num_replicas_in_sync
train_ds = strategy.distribute_datasets_from_function(lambda _: create_train_ds(train_raw,per_sample_batch_size))
valid_ds = strategy.distribute_datasets_from_function(lambda _: create_valid_ds(valid_raw,per_sample_batch_size))
tuner.search(train_data=train_ds,validation_data=valid_ds,epochs=10,batch_size=BATCH_SIZE,steps_per_epoch=steps_per_epoch,validation_steps=validation_steps)
```

**Comment by ajelenak** [comment]
Your problem is from trying to create a file in an object store. Object stores do not (yet) support partial writing to one of their objects. I suggest looking into fsspec [remote write caching](https://filesystem-spec.readthedocs.io/en/latest/features.html#remote-write-caching) as a possible solution.

You can close this issue as it is not an h5py problem.

**Comment by h4ck4l1** [comment]
Hey @ajelenak 

Thanks for responding right away very appreciated.
I am a beginner so let me understand this.
So you mean to say let keras tuner write the files remotely then send them to gs://my_bucket?.
and can you tell me where the with context should be opened? for tuner initiation right? like this?

```python
with fsspec.open("simplecache::gs://my_bucket/my_directory","wb",s3={"profile":"writer"}) as f:
    tuner = kt.Hyperband(
        objective=kt.Objective("Validation Accuracy","max"),
        max_epochs=100,
        factor=2,
        hypermodel=CustomHypermodel(),
        hyperband_iterations=2,
        directory=f,
        project_name="mnist_hyperband",
        overwrite=True
)

```

Is this what you meant?

Last one thing and I will close this issue. How did you recognise its a problem that is not yet supported? supported by who? keras tuner or gcp?
Thankyou very much.

**Comment by ajelenak** [comment]
The quickest fix guaranteed to work is to use a local directory instead of a gcs bucket. Then upload all the created files from that local directory into the bucket once all the processing has finished.

If that is not possible or you prefer the bucket option then it depends on all the packages you use, like Keras, whether they can pass around and work correctly with fsspec objects. H5py will do the right thing if it receives an fsspec object that represents a file but not a remote directory.

**Comment by ajelenak** [comment]
> How did you recognise its a problem that is not yet supported? supported by who? keras tuner or gcp?

It is not supported by cloud object store providers, like AWS, Google, Microsoft, etc. You can either put entire object (file) or replace existing object (file) with a new version. You cannot partially write to an object like you partially write to a file in a file system.

**Comment by h4ck4l1** [comment]
Again thankyou for your swift replies and I got to know a new thing that the problem lies with gcp itself, and also answers my questions as to why no one has an example notebook to run keras_tuner with TPU's on colab. 

I want to save them directly to gcp is because usually the runtimes in colab and kaggle have a timeout limit. If hit, all the files on the system are flushed clean. So I want them directly saved to cloud.

Thanks bro for sharing info...I got to know where the actual problem lies....now time to find alternatives :)
